oryx-app-mllib/src/main/java/com/cloudera/oryx/app/mllib/kmeans/Evaluation.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2015, Cloudera and Intel, Inc. All Rights Reserved.
+ *
+ * Cloudera, Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"). You may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for
+ * the specific language governing permissions and limitations under the
+ * License.
+ */
+
+package com.cloudera.oryx.app.mllib.kmeans;
+
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.api.java.function.PairFunction;
+import org.apache.spark.mllib.linalg.Vector;
+import scala.Tuple2;
+
+import com.cloudera.oryx.app.kmeans.ClusterInfo;
+import com.cloudera.oryx.app.kmeans.SquaredDistanceFn;
+
+final class Evaluation {
+
+  private Evaluation() {
+  }
+
+  /**
+   * Computes the Davies-Bouldin Index, See http://en.wikipedia.org/wiki/Davies-Bouldin_index
+   *
+   * @param clusters - List of {@link ClusterInfo}
+   * @param testData - test data for evaluation
+   * @return Davies Bouldin index - measure of clustering quality, lower the better
+   */
+  static double daviesBouldinIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+    Map<Integer, Tuple2<Double, Long>> clusterSumDistanceAndCounts =
+        fetchClusterSumDistanceAndCounts(clusters, testData).collectAsMap();
+
+    return calcDBIndex(clusterSumDistanceAndCounts, clusters);
+  }
+
+  /**
+   * Computes the Dunn Index of a given clustering, See http://en.wikipedia.org/wiki/Dunn_index
+   * @param clusters - List of {@link ClusterInfo}
+   * @param testData - test data for evaluation
+   * @return Dunn Index - higher the better
+   */
+  static double dunnIndex(final List<ClusterInfo> clusters, JavaRDD<Vector> testData) {
+    List<Tuple2<Integer, Tuple2<Double, Long>>> clusterSumDistanceAndCounts =
+        fetchClusterSumDistanceAndCounts(clusters, testData).collect();
+
+    return calcDunnIndex(clusterSumDistanceAndCounts, clusters);
+  }
+
+  private static double calcDBIndex(Map<Integer, Tuple2<Double, Long>> clusterSumDistAndCounts,
+                                    List<ClusterInfo> clusters) {
+    double totalDBIndex = 0.0;
+    for (int i = 0; i < clusters.size(); i++) {
+      double maxDBIndex = 0;
+      for (int j = 0; j < clusters.size(); j++) {
+        if (i != j) {
+          ClusterInfo c1 = clusters.get(i);
```

Comment:
This variable can be declared outside the inner loop, as well as scatter, and I think the `containsKey(i)` condition can be checked outside too? to avoid whole loops for `i` that won't do anything.